src/Relation/Binary/Rewriting.agda

```diff
@@ -0,0 +1,135 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Concepts from rewriting theory
+-- Definitions are based on "Term Rewriting Systems" by J.W. Klop
+------------------------------------------------------------------------
+
+{-# OPTIONS --with-K --safe #-}
+
+module Relation.Binary.Rewriting where
+
+open import Agda.Builtin.Equality using (_≡_ ; refl)
+open import Data.Product using (_×_ ; ∃ ; _,_ ; proj₂)
+open import Data.Empty
+open import Data.Sum as Sum using (_⊎_)
+open import Level
+open import Relation.Binary.Core
+open import Relation.Binary.Construct.Closure.Transitive
+open import Relation.Binary.Construct.Closure.ReflexiveTransitive
+open import Relation.Binary.Construct.Closure.Symmetric
+open import Relation.Nullary
+
+-- The following definitions are taken from Klop [5]
+IsNormalForm : ∀ {A : Set} → {ℓ : Level} → {r : Rel A ℓ} → (a : A) → Set _
+IsNormalForm {A} {ℓ} {_⟶_} a = ¬ ∃ λ b → (a ⟶ b)
+
+HasNormalForm : ∀ {A : Set} → {ℓ : Level} → {r : Rel A ℓ} → (b : A) → Set _
+HasNormalForm {A} {ℓ} {_⟶_} b = ∃ λ a → ( IsNormalForm {A} {ℓ} {_⟶_} a × (b —↠ a))
+  where
+    _—↠_ = Star _⟶_
+
+NormalForm : ∀ {A : Set} → {ℓ : Level} → (r : Rel A ℓ) → Set _
+NormalForm {A} {ℓ} r = ∀ {a b}
+  → IsNormalForm {A} {ℓ} {r} a
+  → b ↔ a
+  → b —↠ a
+  where
+    _—↠_ = Star r
+    _↔_  = Star (SymClosure r)
+
+WeaklyNormalizing : ∀ {A : Set} → {ℓ : Level} → (r : Rel A ℓ) → Set _
+WeaklyNormalizing {A} {ℓ} r = ∀ a → HasNormalForm {A} {ℓ} {r} a
+
+UniqueNormalForm : ∀ {A : Set} → {ℓ : Level} → (r : Rel A ℓ) → Set _
+UniqueNormalForm {A} {ℓ} r = ∀ {a b}
+  → IsNormalForm {A} {ℓ} {r} a
+  → IsNormalForm {A} {ℓ} {r} b
+  → a ↔ b
+  → a ≡ b
+  where
+    _↔_ = Star (SymClosure r)
+
+Det : ∀ {a b ℓ₁ ℓ₂} → {A : Set a} → {B : Set b} → Rel B ℓ₁ → REL A B ℓ₂ → Set _
+Det _≈_ _—→_ = ∀ {x y z} → x —→ y → x —→ z → y ≈ z
+
+Deterministic : ∀ {a b ℓ} → {A : Set a} → {B : Set b} → REL A B ℓ → Set _
+Deterministic = Det _≡_
```

Comment:
I'm not sure this shorthand follows conventions elsewhere in the standard library. I think it's clearer to rename `Det` to `Deterministic` and just write  `Deterministic _≡_ _—→_` out in full.